torch/distributed/fsdp/shard_utils.py

```diff
@@ -0,0 +1,144 @@
+    placement is not used.
+    """
+    offsets: List[int] = []
+    if isinstance(sharding_spec, EnumerableShardingSpec):
+        for shard in sharding_spec.shards:
+            offsets.append(shard.shard_offsets[0])
+    elif isinstance(sharding_spec, ChunkShardingSpec):
+        assert sharding_spec.dim == 0
+        chunk_size = math.ceil(tensor_numel / world_size)
+        if chunk_size == 1:
+            offsets = [
+                rank if rank < tensor_numel else tensor_numel
+                for rank in range(world_size)
+            ]
+        else:
+            offsets = [chunk_size if rank > 0 else 0 for rank in range(world_size)]
+            offsets = list(itertools.accumulate(offsets))
+    else:
+        raise ValueError(f"Un-recognized sharding spec type {type(sharding_spec)}.")
+
+    return offsets
+
+
+def _offsets_to_split_sizes(
+    input_offsets: List[int],
+    output_offsets: List[int],
+    tensor_numel: int,
+    world_size: int,
+    my_rank: int,
+) -> Tuple[List[int], List[int]]:
+    r"""
+    Given the shard offsets for each rank of the input tensor and output tensor,
+    this API returns the corresponding split sizes that can be passed to
+    all_to_all_single().
+    """
+
+    def _get_interval(offsets):
+        if my_rank != world_size - 1:
+            return offsets[my_rank], offsets[my_rank + 1] - 1
+        else:
+            return offsets[my_rank], tensor_numel - 1
+
+    def _offsets_to_sizes(offsets, begin, end):
+        sizes = []
+        for i, offset in enumerate(offsets):
+            next_offset = offsets[i + 1] if i < len(offsets) - 1 else end + 1
+            sizes.append(
+                (next_offset - offset)
+                - max(begin - offset, 0)
+                - max(next_offset - end - 1, 0)
+            )
+        return sizes
+
+    def _convert(from_offsets, to_offsets, split_sizes):
+        begin, end = _get_interval(from_offsets)
+        to_begin_rank = bisect.bisect(to_offsets, begin) - 1
+        to_end_rank = bisect.bisect(to_offsets, end) - 1
+        _split_sizes = _offsets_to_sizes(
+            to_offsets[to_begin_rank : to_end_rank + 1], begin, end
+        )
+        split_sizes[to_begin_rank : to_end_rank + 1] = _split_sizes
+
+    input_split_sizes = [0 for _ in range(world_size)]
+    output_split_sizes = [0 for _ in range(world_size)]
+    _convert(input_offsets, output_offsets, input_split_sizes)
+    _convert(output_offsets, input_offsets, output_split_sizes)
+
+    return input_split_sizes, output_split_sizes
+
+
+def reshard_flatten_tensor(
+    input_tensor: ShardedTensor,
+    output_spec: ShardingSpec,
+    world_size: int,
+    my_rank: int,
+    device: torch.device,
+    process_group: Optional[ProcessGroup],
+) -> torch.Tensor:
+    """
+    Resharded a sharded flatten tensor, this is used by FSDP to do sharded
+    state_dict. But the functionaility is not supported by ShardedTensor.
```

Comment:
Should we eventually generalize + upstream `reshard_flatten_tensor` if it is useful more generally in sharding package? cc @pritamdamania87 @wanchaol

Reply:
I discussed with @fduwjj , if reshard_flatten_tensor is general, we can upstream it to the sharding package. Right now, there are some gaps.